Add SPARC architecture

yashi opened the pull request:
This PR is to add SPARC architecture to Zephyr RTOS.  This PR is only for QEMU target.

It's my first PR to add new architecture to Zephyr.  Please let me know if there is any procedure I should follow to.

This PR needs #16950 to be merged before.

Hoping to be merged for version 2.0.

yashi committed:
arch: sparc: Initial commit

This is the initial commit to add SPARC architecture.  This commit is
very basic and adds very minimum code to run `sample/hello_world` on
SPARC QEMU.  No UART, no timer, no interrupt, and no threads support
is included.  We tried to keep this commit small as possible so that
people can easily review what we are tring to add.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>
Signed-off-by: Daisuke Motooka <d-motooka@ispace-inc.com>
Signed-off-by: Itsuki Kamino <i-kamino@ispace-inc.com>

yashi committed:
drivers: serial: Add SPARC LEON2 UART driver

Add UART driver for SPARC LEON2.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>
Signed-off-by: Daisuke Motooka <d-motooka@ispace-inc.com>

yashi committed:
arch: sparc: Enable ATOMIC_OPERATIONS_C

SPARC V8 doesn't have native atomic ops.  Use the pure C version of
functions, instead.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

yashi committed:
boards: sparc: qemu_at697: Enable timer

CONFIG_SYS_CLOCK_TICKS_PER_SEC must be non-zero to build timer
drivers.  And select LEON2 timer driver for QEMU AT697.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

yashi committed:
boards: sparc: qemu_at697: Enable multi-threading

Enable multi-threading for QEMU SPARC AT697.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

Commit:
drivers: timer: Add SPARC LEON2 timer driver

Add timer driver for SPARC LEON2.

Signed-off-by: Daisuke Motooka <d-motooka@ispace-inc.com>
Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

yashi committed:
arch: sparc: Add Interrupt support

Add interrupt handling support for SPARC.  This commit adds both LEON2
architecture code and a device driver for AT697 interrupt controller,
because they depends each other.

This implementation is based on auto generated isr table.

SPARC "ldd" instruction, used to load both the corresponding ISR and
the argument, requires the given address to be aligned to 8 byte
boundary.  Thus, we set ARCH_SW_ISR_TABLE_ALIGN to 8.

This version still assumes single thread mode.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>
Signed-off-by: Daisuke Motooka <d-motooka@ispace-inc.com>

yashi committed:
arch: sparc: Add cooperative multi-threading support

This commit add cooperative multi-threading support to SPARC
architecture.

Because LEON2 doesn't have a "wait for interrupt" instruction, __weak
is added for arch_cpu_idle() to allow a board or a SOC to override the
function.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>
Signed-off-by: Daisuke Motooka <d-motooka@ispace-inc.com>

yashi committed:
arch: sparc: Add preemptive scheduling support

This commit adds preemptive scheduling to SPARC architecture.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>
Signed-off-by: Daisuke Motooka <d-motooka@ispace-inc.com>

yashi committed:
boards: sparc: qemu_at697: Enable preemptive scheduling

Enable preemptive scheduling for QEMU SPARC AT697.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

yashi committed:
arch: sparc: Add irq offload support

This commit add irq_offload() function support to SPARC architecture.
irq_offload() is basically for test cases but is used extensively in
many test cases in Zephyr.  Without it, not much test cases pass.

We use the last software trap number, 0xff, for irq_offload().

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

yashi committed:
arch: sparc: clock: Use prescaler

LEON2 timer counter is only 24 bits.  The bit length along with system
clock of 50 MHz is not sufficient to be used for 500 ms busy wait
which we do in tests/kernel/common/.

Use prescaler to make 4 times longer.  This yields 1340 ms or 1.34 sec
duration and 80 ns resolution, instead of the original 335 ms duration
and 20 ns resolution.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

yashi committed:
arch: sparc: Move irq definitions to a public header

Some tests under tests/kernel/context need to be able to enable /
disable timer interrupt.  Thus, the interrupt number definition for
LEON2 timer unit needs to be in a public header.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

yashi committed:
arch: sparc: Add an accessor for PSR

SPARC's Processor State Register contains many important fields.  One
of those is ET (Enable Traps) bit.  The bit indicates whether the CPU
accepts a trap or not.  This accessor is basically for irq_offload()
since usual programs won't need to worry about ET bit.

irq_offload() uses "ta" instruction to make a software interrupt but
if ET bit is 0, we won't get trapped as we intended.  Thus we have
assertion to check that trap is enabled.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

yashi committed:
arch: sparc: Add stack sentinel support

This commit adds stack sentinel support for SPARC architecture.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

yashi committed:
arch: sparc: Enlarge the stack size for sanity check

SPARC architecture needs much more stack size than others.  2048 is
the minimum for sanity check to pass.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

yashi committed:
boards: sparc: qemu_at697: Add extra stack size

SPARC archtecture needs more stack when we run sanity check.  Increase
another 2048 bytes for testing.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

yashi committed:
tests: kernel: context: Fix SPARC faiulre

This commit fixes kernel context tests.

 - Define TICK_IRQ needed in tests/kernel/context.
 - Skip k_cpu_idle() check since SPARC don't have power-saving
   instruction.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

SebastianBoe commented on cmake/compiler/gcc/target.cmake:
```diff
@@ -80,12 +84,16 @@ if(NOT no_libgcc)
-  LIST(APPEND TOOLCHAIN_LIBS gcc)
+    LIST(APPEND LIB_INCLUDE_DIR "-L\"${LIBGCC_DIR}\"")
+    LIST(APPEND TOOLCHAIN_LIBS gcc)
+    endif()
```
indentation

yashi replied:
fixed.

SebastianBoe commented on cmake/compiler/gcc/target.cmake:
```diff
@@ -68,6 +68,10 @@ elseif("${ARCH}" STREQUAL "arc")
     )
+elseif("${ARCH}" STREQUAL "sparc")
+  list(APPEND TOOLCHAIN_C_FLAGS
+    -mcpu=leon -mflat
```
Don't add -mflat twice.

yashi replied:
I've moved `-mflat` from `arch/sparc/CMakeLists.txt` and `cmake/compiler/gcc/target.cmake` to `cmake/compiler/gcc/target_sparc.cmake` like arm and x86.

SebastianBoe commented on cmake/compiler/gcc/target.cmake:
```diff
@@ -80,12 +84,16 @@ if(NOT no_libgcc)
 
-  get_filename_component(LIBGCC_DIR ${LIBGCC_FILE_NAME} DIRECTORY)
+  if (CONFIG_SPARC)
+    LIST(APPEND TOOLCHAIN_LIBS ${LIBGCC_FILE_NAME})
```
Either add an explanation to why this needs to be different with SPARC as a comment in the code, or preferably, align with the other architecture's and avoid this patch.

yashi replied:
We'll postpone this change after sparc is merged.  Because, for now, we are targeting GCC in SDK, we don't need this.   I'll later open another issue just for this.

SebastianBoe commented on drivers/timer/Kconfig:
```diff
@@ -192,6 +192,14 @@ config LITEX_TIMER
 
+config LEON2_TIMER
+	bool "LEON2 Timer Unit"
+	default n
```
It is best-practice to omit default n.

yashi replied:
done

SebastianBoe commented on soc/sparc/at697/CMakeLists.txt:
```diff
@@ -0,0 +1 @@
+zephyr_compile_options(-mcpu=leon -mtune=leon)
```
Find a way to avoid specifying -mcpu=leon multiple times.

yashi replied:
In the current version, we only have "-mcpu=leon -mtune=leon" only in this file.

andrewboie commented on arch/Kconfig:
```diff
@@ -46,6 +46,13 @@ config RISCV32
+	select BIG_ENDIAN
+	select USE_SWITCH_SUPPORTED
+	select USE_SWITCH
+	select ATOMIC_OPERATIONS_C
```
why? shouldn't you be using GCC builtins?

yashi replied:
Current build doesn't link against libatomic and failed with the following error:
```
include/sys/atomic.h:223: undefined reference to `__atomic_fetch_or_4'
include/sys/atomic.h:270: undefined reference to `__atomic_fetch_and_4'
```

In fact, many arch currently selects pure C version of it.   Would it be OK with you if I say I'll create an issue to convert to builtins once this PR is merged?

andrewboie replied:
toolchain may not implement them, we had this issue with Nios II. fine to use C for now

andrewboie commented on arch/sparc/core/cpu_idle.c:
```diff
@@ -0,0 +1,16 @@
+	 * but the unlock key is lost. so unlock irq unconditionally
+	 * irq_unlock function will enable unlock interrupt if key has
+	 * Enable Trap bit on.*/
+	irq_unlock(0);
```
this function will immediately return - does SPARC have a "wait for interrupt" instruction because that's what this function needs to do

yashi replied:
Unfortunately it doesn't have the instruction.  Instead, it's a implementation dependent.  Some LEON3 has it in Ancillary State Register:

> 2.2.13 Power-down
> The processor can be configured to include a power-down feature to minimize power consumption during idle periods. The power-down mode is entered by performing a WRASR instruction to %asr19:
> ```
> wr %g0, %asr19
> ```
> During power-down, the pipeline is halted until the next interrupt occurs. Signals inside the processor pipeline and caches are then static, reducing power consumption from dynamic switching.

But this PR is for LEON **2**, which doesn't have it, neither.

andrewboie replied:
OK

andrewboie replied:
can you declare this __weak so it may be overridden in board or SOC configuration?

yashi replied:
Added __weak to `arch_cpu_idle()` and comment about it in the commit log.

andrewboie commented on arch/sparc/core/entry.S:
```diff
@@ -0,0 +1,282 @@
+GTEXT(__start)
+
+/* imports */
+GTEXT(_PrepC)
```
you need to namespace all your private variables and functions to this arch port with z_, preferably z_sparc_ for clarity

yashi replied:
done.

galak reviewed (changes requested):
This needs an update to sdk-ng to build a toolchain for SPARC.

yashi committed:
CODEOWNERS: Add owner for SPARC architecture

Add myself to SPARC code owner.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

tyrantlizard77 reviewed (dismissed):
LGTM once SDK and CI issues resolved

yashi commented:
Thanks.  Regarding to the CI issues, I mean the warnings and errors `checkpatch`, I think I've fixed I can.  The remains are:

- WARNING:NEW_TYPEDEFS: do not add new typedefs
  - I think I need typedefs because these are used by `scripts/gen_offset_header.py`
- ERROR:MULTISTATEMENT_MACRO_USE_DO_WHILE: Macros with multiple statements should be enclosed in a do - while loop
  - I don't think we can wrap linker directives with do-while.
- typedef off_t
  - need to have it.

What should I do about them?

andrewboie commented:
Ah right, checkpatch errors don't hold up merge, whoever merges will take them into account. Shippable is happy, so if we can get a few more reviews in the next few days we should be able to make for 2.0

Is this arch currently being built by CI?

yashi commented:
No.  Because the SDK don't have a SPARC compiler, I don't think the CI is building it. I'll take a look at sdk-ng after I create a separate issue about libgcc and compiler flags @SebastianBoe commented.

galak commented:
> No. Because the SDK don't have a SPARC compiler, I don't think the CI is building it. I'll take a look at sdk-ng after I create a separate issue about libgcc and compiler flags @SebastianBoe commented.

What toolsuite are you using to develop this?

yashi commented:
I've been using GCC from https://www.gaisler.com/index.php/downloads/compilers .
I've tried `gcc-8-sparc64-linux-gnu` from Debian but haven't successfully built; always get `SPARC32PLUS, V8+ Required` set in the final binary, even with `-mno-v8plus`.

yashi commented:
Just in case anyone is interested in.  A diff of `gcc -Q --help=target`:
```diff
--- /tmp/debian.txt	2019-08-07 19:45:25.033387977 +0900
+++ /tmp/bcc.txt	2019-08-07 19:45:30.637429500 +0900
@@ -1,40 +1,36 @@
 
-  -m32                        		[disabled]
-  -m64                        		[enabled]
+  -m32                        		[enabled]
+  -m64                        		[disabled]
   -mapp-regs                  		[enabled]
-  -mbionic                    		[disabled]
   -mcbcond                    		[disabled]
   -mcmodel=                   		
-  -mcpu=                      		v9
+  -mcpu=                      		v7
   -mdebug=                    		
   -mfaster-structs            		[disabled]
   -mfix-at697f                		[disabled]
+  -mfix-b2bst                 		[disabled]
   -mfix-gr712rc               		[disabled]
+  -mfix-tn0013                		[disabled]
   -mfix-ut699                 		[disabled]
   -mfix-ut700                 		[disabled]
   -mflat                      		[disabled]
   -mfmaf                      		[disabled]
   -mfpu                       		[enabled]
-  -mfsmuld                    		[enabled]
-  -mglibc                     		[enabled]
+  -mfsmuld                    		[disabled]
   -mhard-float                		[enabled]
   -mhard-quad-float           		[disabled]
-  -mlong-double-128           		[enabled]
-  -mlong-double-64            		[disabled]
-  -mlra                       		[enabled]
-  -mmemory-model=             		tso
-  -mmusl                      		[disabled]
+  -mlra                       		[disabled]
+  -mmemory-model=             		default
   -mpopc                      		[disabled]
-  -mptr32                     		[disabled]
-  -mptr64                     		[enabled]
+  -mptr32                     		[enabled]
+  -mptr64                     		[disabled]
   -mrelax                     		[disabled]
   -msoft-float                		[disabled]
   -msoft-quad-float           		[enabled]
-  -mstack-bias                		[enabled]
+  -mstack-bias                		[disabled]
   -mstd-struct-return         		[disabled]
   -msubxc                     		[disabled]
-  -mtune=                     		v9
-  -muclibc                    		[disabled]
+  -mtune=                     		v7
   -munaligned-doubles         		[disabled]
   -muser-mode                 		[enabled]
   -mv8plus                    		[disabled]
```

galak commented:
> I've been using GCC from [gaisler.com/index.php/downloads/compilers](https://www.gaisler.com/index.php/downloads/compilers) .
> I've tried `gcc-8-sparc64-linux-gnu` from Debian but haven't successfully built; always get `SPARC32PLUS, V8+ Required` set in the final binary, even with `-mno-v8plus`.

and which qemu?

yashi commented:
For QEMU, we've been using https://github.com/AdaCore/qemu with a few fixes.

galak commented:
I know you hoped to get this into 2.0.0, but the cut off for that is this Friday.  Since we don't get have the SDK support needed for this it seems highly unlikely we'll be able to get this in for 2.0.0.  However I don't see any reason that we can't get this in for 2.1.0

yashi commented:
@galak, I agree.  I thought it'd be a nice addition (a new arch) for a major version bump.  But, it's just that.  No rush from me to squeeze it into 2.0.  Thank you all, as always, for the kind support.  I'll finish other things (preparing sdk and the libgcc issue) first and then come back to this after the 2.0 release.

yashi committed:
sparc: Handle orphaned sections

Newer GCC emits warning about orphaned sections.  Suppress the
warnings by adding the sections where they should be.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

andrewboie commented on arch/sparc/core/irq_offload.c:
```diff
@@ -0,0 +1,38 @@
+	__ASSERT((z_sparc_get_psr() & PSR_ET) != 0,
+		 "%s called with trap disabled\n", __func__);
+
+	k_sched_lock();
```
don't need this anymore once #20397 lands (very soon)

yashi replied:
Removed sched lock.

tyrantlizard77 reviewed (approved):
some minor suggestions, otherwise looks great
please address @SebastianBoe 's comments

yashi commented:
Took me a while to came back to this.  I've rebased to the current master.   Addressed a few places pointed by reviews.  I haven't fixed points where @SebastianBoe pointed out, just yet.  I'll work on them next.  This push is to see how everything else goes.

andrewboie commented:
Where are we at with getting a SPARC toolchain/emulator in the SDK?

stephanosio commented:
> Where are we at with getting a SPARC toolchain/emulator in the SDK?

Both toolchain and emulator (qemu) were added in Zephyr SDK 0.11.0 Alpha 7, which will be available in the CI image once #20625 is merged.

yashi commented:
> This needs an update to sdk-ng to build a toolchain for SPARC.

It has been merged.
https://github.com/zephyrproject-rtos/sdk-ng/pull/148

yashi commented:
Forgot to push LEON2 patches for QEMU.  The current patch is just to enable SPARC arch.  I'll address it in sdk-ng.

yashi commented:
I've addressed everything reviewers pointed out.  Thanks all.  Let me know if there still is anything I can fix.

yashi commented:
Rebased on v2.1.

andrewboie commented:
Needs a rebase.

@galak the new SDK is in place, can you re-visit your review?

yashi committed:
arch: sparc: Add memory bit manipulation functions

This commit adds bit manipulation functions at a memory location.  The
functions are all based on sys_read* and sys_write* functions and are
implemented in C.

There are C implementations in include/arch/common/sys_io.h, but they:

 - only support little endian.
 - are not using sys_read/write.

It sure is better to have endian aware implementation in common
place.  But this is a start.

Signed-off-by: Yasushi SHOJI <y-shoji@ispace-inc.com>

yashi commented:
I've rebased to the current master.  It compiles and runs `samples/philosophers` just fine.

Unfortunately, many tests by `sanitycheck` fail.  What is the best way to disable unsupported test in `boards/*/*/*.yaml`?  I was gonna use `ignore_tags:` but it's kind of tedious to list them all.  Is there better way?

yashi commented:
Hmmm it seems it's qemu.  I'll take a look.
